platforms/allwinner-d1/core/src/drivers/sharp_display.rs

```diff
@@ -390,82 +346,51 @@ struct Context {
 }
 
 /// Waiter for "changes have been made to the working frame buffer"
-static DIRTY: WaitCell = WaitCell::new();
+static DIRTY: WaitQueue = WaitQueue::new();
 
-/// Create and return a Simulator display object from raw pixel data.
-///
-/// Pixel data is turned into a raw image, and then drawn onto a SimulatorDisplay object
-/// This is necessary as a e-g Window only accepts SimulatorDisplay object
-/// On a physical display, the raw pixel data can be sent over to the display directly
-/// Using the display's device interface
-fn frame_display(fc: &mut FrameChunk) -> Result<ImageRaw<Gray8>, ()> {
-    let raw_image: ImageRaw<Gray8>;
-    raw_image = ImageRaw::<Gray8>::new(fc.bytes.as_slice(), fc.width);
-    Ok(raw_image)
-}
+fn draw_to(dest: &mut FullFrame, src: &MonoChunk, width: u32, height: u32) {
+    let meta = src.meta();
+    let data = src.data();
+    let mask = src.mask();
 
-struct FrameInfo {
-    frame: u16,
-}
+    let start_x = meta.start_x();
+    let start_y = meta.start_y();
+    let src_width = meta.width();
 
-struct DisplayInfo {
-    frame_idx: u16,
-    frames: FixedVec<FrameInfo>,
-}
+    if start_y >= height {
+        return;
+    }
+    if start_x >= width {
+        return;
+    }
 
-impl DisplayInfo {
-    // Returns a new frame chunk
-    async fn new_frame(
-        &mut self,
-        start_x: i32,
-        start_y: i32,
-        width: u32,
-        height: u32,
-    ) -> Result<FrameChunk, FrameError> {
-        let fidx = self.frame_idx;
-        self.frame_idx = self.frame_idx.wrapping_add(1);
-
-        self.frames
-            .try_push(FrameInfo { frame: fidx })
-            .map_err(|_| FrameError::NoFrameAvailable)?;
-
-        let size = (width * height) as usize;
-
-        // TODO: So, in the future, we might not want to ACTUALLY allocate here. Instead,
-        // we might want to allocate ALL potential frame chunks at registration time and
-        // hand those out, rather than requiring an allocation here.
-        //
-        // TODO: We might want to do ANY input checking here:
-        //
-        // * Making sure the request is smaller than the actual display
-        // * Making sure the request exists entirely within the actual display
-        let mut bytes = FixedVec::new(size).await;
-        for _ in 0..size {
-            let _ = bytes.try_push(0);
-        }
-        let fc = FrameChunk {
-            frame_id: fidx,
-            bytes,
-            start_x,
-            start_y,
-            width,
-            height,
-        };
+    let s = data
+        .chunks(src_width as usize)
+        .zip(mask.chunks(src_width as usize));
 
-        Ok(fc)
-    }
+    let before = dest.dirty_line.iter().filter(|b| **b).count();
```

Comment:
nit, take it or leave it: it might be worth guarding this behind [`event_enabled!`](https://docs.rs/tracing/latest/tracing/macro.event_enabled.html), since it's only used if `trace!` is enabled, and we could avoid iterating if we're not going to log the trace event?
```suggestion
    let before = if tracing::event_enabled!(tracing::Level::TRACE) {
         dest.dirty_line.iter().filter(|b| **b).count()
    } else {
        0
    };
``` 
this is probably a premature optimization though :woman_shrugging:

Reply:
Yeah, I'd probably pass since iterating over bools is probably going to optimize extremely well.